src/parse.js

```diff
@@ -0,0 +1,325 @@
+module.exports = makeParse;
+function makeParse(charIndexBuggy, fromCharCode, newSyntaxError, getClass, forOwn) {
+  // Public: `JSON.parse` parses a JSON source string. See
+  // ES 5.1 section 15.12.2.
+  function parse(source, callback) {
+    Index = 0;
+    Source = "" + source;
+    var firstToken = lex(charIndexBuggy, fromCharCode, newSyntaxError);
+    var result = get(charIndexBuggy, fromCharCode, newSyntaxError, firstToken);
+    // If a JSON string contains multiple tokens, it is invalid.
+    if (lex(charIndexBuggy, fromCharCode, newSyntaxError) != "$") {
+      abort(newSyntaxError);
+    }
+    // Reset the parser state.
+    Index = Source = null;
+    if (callback && getClass.call(callback) == "[object Function]") {
+      var value = {};
+      value[""] = result;
+      return walk(getClass, forOwn, value, "", callback);
+    }
+    return result;
+  }
+  return parse;
+}
+
+// Internal: A map of escaped control characters and their unescaped
+// equivalents.
+var Unescapes = {
+  92: "\\",
+  34: '"',
+  47: "/",
+  98: "\b",
+  116: "\t",
+  110: "\n",
+  102: "\f",
+  114: "\r"
+};
+
+// Internal: Stores the parser state.
+var Index, Source;
+
+// Internal: Resets the parser state and throws a `newSyntaxError`.
+function abort(newSyntaxError) {
+  Index = Source = null;
+  throw newSyntaxError();
+}
+
+// Internal: Returns the next token, or `"$"` if the parser has reached
+// the end of the source string. A token may be a string, number, `null`
+// literal, or Boolean literal.
+function lex(charIndexBuggy, fromCharCode, newSyntaxError) {
```

Comment:
Passing environment dependencies around gets old quickly, but we can't use a global `environment` module that provides these functions because of `runInContext`.